### PR TITLE
A basic Helm chart

### PR DESCRIPTION
**Pull request by yaronius**
Hi there! Created a basic Helm chart, tested locally with minikube. 
Should close https://github.com/appleboy/gorush/issues/469

### CODE REVIEW

**Comment by yaronius** [comment]
I guess those tests didn't fail because of a couple of new YAML files 🤞🏻 🤔

File: helm/gorush/templates/tests/test-connection.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    {{- include "gorush.labels" . | nindent 4 }}
+  annotations:
+    "helm.sh/hook": test
+spec:
+  containers:
+    - name: wget
+      image: busybox
+      command: ['wget']
+      args: ['{{ include "gorush.fullname" . }}:{{ .Values.service.port }}']
+  restartPolicy: Never
```
Comment (chaosaffe) [inline helm/gorush/templates/tests/test-connection.yaml:15]: This should be handled by the health and ready checks on the pods. If you wanted an example consumer that should be a separate chart
```suggestion
```

File: helm/gorush/templates/redis-deployment.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      labels:
+        app: redis
+        role: master
+        tier: backend
+    spec:
+      containers:
+        - name: master
+          image: redis
+          ports:
+            - containerPort: {{ .Values.redis.port }}
```
Comment (chaosaffe) [inline helm/gorush/templates/redis-deployment.yml:26]: Please use an existing [Redis chart](https://github.com/bitnami/charts/tree/master/bitnami/redis) and include it as an optional dependency. 

As gorush doesn't mandate that we use Redis (we can use many backends) this needs to be optional for the user to enable.

```suggestion
```

File: helm/gorush/templates/redis-service.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  labels:
+    {{- include "gorush.labels" . | nindent 4 }}
+spec:
+  ports:
+    - port: {{ .Values.redis.port }}
+      targetPort: {{ .Values.redis.port }}
+  selector:
+    app: redis
+    role: master
+    tier: backend
```
Comment (chaosaffe) [inline helm/gorush/templates/redis-service.yml:15]: As above
```suggestion
```

File: helm/gorush/values.yaml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  #   cpu: 100m
+  #   memory: 128Mi
+  # requests:
+  #   cpu: 100m
+  #   memory: 128Mi
+
+autoscaling:
+  enabled: false
+  minReplicas: 1
+  maxReplicas: 100
```
Comment (chaosaffe) [inline helm/gorush/values.yaml:74]: Let's provide the user with some sane defaults as guard rails

```suggestion
  maxReplicas: 10
```

**Review by chaosaffe — CHANGES_REQUESTED** [review]
Thank you @yaronius for contributing this chart!

Overall it looks great, just needs a few nits cleaned up to meet best practices

File: helm/gorush/templates/redis-deployment.yml (hunk 2 above)
Comment (yaronius) [reply helm/gorush/templates/redis-deployment.yml]: fixed

File: helm/gorush/templates/redis-service.yml (hunk 3 above)
Comment (yaronius) [reply helm/gorush/templates/redis-service.yml]: fixed

**Comment by yaronius** [comment]
Builds are now (almost) passing, the last ones failed because of this error 
```
toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading
```

File: gorush/notification_apns.go [hunk 5]
```diff
@@ -159,11 +159,11 @@ func newApnsClient(certificate tls.Certificate) (*apns2.Client, error) {
 		IdleConnTimeout: idleConnTimeout,
 	}
 
-	if h2Transport, err := http2.ConfigureTransports(transport); err != nil {
+	h2Transport, err := http2.ConfigureTransports(transport)
+	if err != nil {
```
Comment (yaronius) [inline gorush/notification_apns.go:163]: This change and below are to fix linter issues that caused build to fail

File: netlify.toml [hunk 6]
```diff
@@ -3,6 +3,7 @@
   functions = "release/linux/lambda"
 
 [build.environment]
+  GO_VERSION = "1.15"
```
Comment (yaronius) [inline netlify.toml:6]: This fixes Netlify builds because their default build image uses 1.12 🤦🏻

**Comment by yaronius** [comment]
@chaosaffe thanks for the feedback, I tried to fix the things you mentioned but I cannot re-request your review for some reason 🤷🏻‍♂️

**Comment by chaosaffe** [comment]
We are also going to need a way to inject secrets and other configuration parameters. I think it might be better for us to source a `config.yaml` and volume mount it into the gorush container, then create secrets for Android, iOS and Huawei params and source them as environment variables

File: helm/gorush/templates/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+              path: /healthz
+              port: http
+            initialDelaySeconds: 3
+            periodSeconds: 3
+          env:
+          - name: GORUSH_STAT_ENGINE
+            valueFrom:
+              configMapKeyRef:
+                name: {{ .Chart.Name }}-config
+                key: stat.engine
```
[the remark this answers is not in the document]
Comment (chaosaffe) [reply helm/gorush/templates/deployment.yaml:50]: Ignore this, I would rather follow the convention in the config file

[the remark this answers is not in the document]
Comment (yaronius) [reply helm/gorush/templates/deployment.yaml:50]: Ok, so I'll mark it resolved

**Comment by chaosaffe** [comment]
Looks really good! I've got a PR to drop after this one lands to add secrets handling for the chart.

**Comment by chaosaffe** [comment]
@appleboy can we get a review or update on this?
@yaronius can you rebase this please?

**Comment by appleboy** [comment]
@yaronius @chaosaffe Thanks for your contribution.

**Pull request merged**